nir/src/main/scala/scala/scalanative/nir/Vals.scala

```diff
@@ -183,18 +262,37 @@ object Val {
       case _ => false
     }
   }
+
+  /** A heterogeneous collection of data members. */
   final case class StructValue(values: Seq[Val]) extends Val
+
+  /** A homogeneous collection of data members. */
   final case class ArrayValue(elemty: nir.Type, values: Seq[Val]) extends Val
+
+  /** A collection of bytes. */
   final case class ByteString(bytes: Array[scala.Byte]) extends Val {
     def byteCount: scala.Int = bytes.length + 1
   }
+
+  /** A local SSA variable. */
   final case class Local(id: nir.Local, valty: nir.Type) extends Val
+
+  /** A symbol. */
   final case class Global(name: nir.Global, valty: nir.Type) extends Val
 
-  // high-level
+  /** The unit value. */
   case object Unit extends Val
+
+  /** A constant. */
   final case class Const(value: Val) extends Val
+
+  /** A character string. */
   final case class String(value: java.lang.String) extends Val
+
+  /** A virtual value. */
```

Comment:
Note: Virtual values exists only in optimizer. They are not serializable, and would never be emitted by compiler plugin.